File: lab2/NotesApp/services/auth-service.js
// src/services/auth-service.js
import { Account, ID } from "appwrite";
import client from "./appwrite-config";

class AuthService {
  constructor() {
    this.account = new Account(client);
  }

  // Create a new user account and automatically log them in
  async createAccount(email, password, name) {
    try {
      const userAccount = await this.account.create(
        ID.unique(),
        email,
        password,
        name
      );

      if (userAccount) {
        // Automatically log in after account creation
        return await this.login(email, password);
      }

      return userAccount;
    } catch (error) {
      console.error("Error creating account:", error);
      throw error;
    }
  }

  // Log in with email and password
  async login(email, password) {
    try {
      const session = await this.account.createEmailPasswordSession(
        email,
        password
      );
      return session;
    } catch (error) {
      console.error("Error logging in:", error);
      throw error;
    }
  }

  // Get the current logged in user
  async getCurrentUser() {
    try {
      const user = await this.account.get();
      return user;
    } catch (error) {
      // User is not logged in
      return null;
    }
  }

  // Log out the current user
  async logout() {
    try {
      await this.account.deleteSession("current");
      return { success: true };
    } catch (error) {
      console.error("Error logging out:", error);
      throw error;
    }
  }
}

// Export a single instance
const authService = new AuthService();
export default authService;

File: lab2/notes_app/lib/services/auth_service.dart
import 'package:appwrite/appwrite.dart';
import 'package:appwrite/models.dart' show Session;
import 'package:appwrite/models.dart' as models;

import 'appwrite_config.dart';

/// Encapsulates all Appwrite authentication flows used in the app.
class AuthService {
  final Client _client = getClient();
  late final Account _account;

  AuthService() {
    _account = Account(_client);
  }

  /// Creates a user account then signs in to return the active session.
  Future<Session> createAccount(
    String email,
    String password,
    String name,
  ) async {
    await _account.create(
      userId: ID.unique(),
      email: email,
      password: password,
      name: name,
    );
    return login(email, password);
  }

  /// Signs the user in with email/password credentials.
  Future<Session> login(String email, String password) {
    return _account.createEmailSession(email: email, password: password);
  }

  /// Returns the currently authenticated user or null when no session exists.
  Future<models.Account?> getCurrentUser() async {
    try {
      return await _account.get();
    } on AppwriteException catch (error) {
      if (error.code == 401) {
        return null;
      }
      rethrow;
    }
  }

  /// Clears the active session on the device.
  Future<void> logout() {
    return _account.deleteSession(sessionId: 'current');
  }
}
